### ISSUE TITLE
Idle timeout exception on Http.singleRequest

### ISSUE THREAD

**Issue by Gabology** [akka/akka-http#1176]
Seems it should be fixed since earlier but I'm running `akka-http_2.11: 10.0.6` and still getting this stack trace:

```
[DEBUG] [06/05/2017 07:06:19.604] [default-akka.actor.default-dispatcher-4898] [akka.stream.Log(akka://default/user/StreamSupervisor-0)] [response] Downstream finished.
[DEBUG] [06/05/2017 07:06:49.613] [default-akka.actor.default-dispatcher-4897] [default/Pool(shared->https://localhost:443)] Shutting down host connection pool
[DEBUG] [06/05/2017 07:06:49.614] [default-akka.actor.default-dispatcher-4901] [default/Pool(shared->https://localhost:443)] [0] </172.17.0.3:44497->localhost:443> Slot was stopped
[DEBUG] [06/05/2017 07:06:49.614] [default-akka.actor.default-dispatcher-4901] [default/Pool(shared->https://localhost:443)] [1] <unconnected> Slot was stopped
[DEBUG] [06/05/2017 07:06:49.614] [default-akka.actor.default-dispatcher-4901] [default/Pool(shared->https://localhost:443)] [2] <unconnected> Slot was stopped
[DEBUG] [06/05/2017 07:06:49.614] [default-akka.actor.default-dispatcher-4901] [default/Pool(shared->https://localhost:443)] [3] <unconnected> Slot was stopped
[DEBUG] [06/05/2017 07:07:26.274] [default-akka.actor.default-dispatcher-4897] [akka://default/user/StreamSupervisor-2/flow-494-0-unknown-operation] Aborting tcp connection to localhost:443 because of upstream failure: TCP idle-timeout encountered on connection to [localhost:443], no bytes passed in the last 1 minute
akka.stream.impl.io.TcpIdleTimeout$$anonfun$1.applyOrElse(TcpStages.scala:374)
akka.stream.impl.io.TcpIdleTimeout$$anonfun$1.applyOrElse(TcpStages.scala:372)
scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
akka.stream.impl.fusing.MapError$$anon$12.onUpstreamFailure(Ops.scala:285)
akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:733)
akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:616)
akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:471)
akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:423)
akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:603)
akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:618)
akka.actor.Actor$class.aroundReceive(Actor.scala:497)
akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:529)
akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
akka.actor.ActorCell.invoke(ActorCell.scala:495)
akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
akka.dispatch.Mailbox.run(Mailbox.scala:224)
akka.dispatch.Mailbox.exec(Mailbox.scala:234)
scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

I'm only using `Http.singleRequest` so nothing low level.

**Comment by ktoso** [comment]
It's logged on DEBUG level as you can see.. 
Yes, idle timeouts use the failure signal of streams to shut down the stream - by design.
Yes idle timeouts are configurable and documented.

What are you expecting?

**Comment by OverlyExcessive** [comment]
Why is it still printing out the stack trace? In my use case I am logging some other valuable debug information so I cannot set the log level to a higher level but this will clutter the logs and the presence of the stack trace indicates that there is some error.

**Comment by jrudolph** [comment]
I guess we could remove the stack trace from the exception as it serves no purpose.

I opened a PR for that here:

https://github.com/akka/akka/pull/23106

**Comment by jrudolph** [comment]
Ticket can be closed once new akka version has been released with the fix and akka-http has been upgraded to use it.

**Comment by ktoso** [comment]
Thanks Johannes, I agree that that stacktrace is not useful there 👍 Thanks for noticing @OverlyExcessive !

**Comment by ktoso** [comment]
Re-opening, this can only be closed once conditions @jrudolph explained are met (we upgrade the dep)

**Comment by sgoyal1012** [comment]
[Related To This Issue]

**Problem+ Stack Trace**
Hi all, I am using the host level connection to make a GET request to a server. My reason to use the host level connection instead of the single level request is to make use of the retry functionality tuned by the 'max-retries' params.

After successfully receiving the response from the server, the client side API further down the line spits out this stack trace: (My idle timeout is set to 50 seconds, as can be seen in the log). 
```
 akka.stream.scaladsl.TcpIdleTimeoutException: TCP idle-timeout encountered on connection to [localhost:3366], no bytes passed in the last 50 seconds
        at akka.stream.impl.io.TcpIdleTimeout$$anonfun$1.applyOrElse(TcpStages.scala:386) ~[codegen.jar:0.1]
        at akka.stream.impl.io.TcpIdleTimeout$$anonfun$1.applyOrElse(TcpStages.scala:384) ~[codegen.jar:0.1]
        at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36) ~[codegen.jar:0.1]
        at akka.stream.impl.fusing.MapError$$anon$11.onUpstreamFailure(Ops.scala:289) ~[codegen.jar:0.1]
        at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:485) ~[codegen.jar:0.1]
        at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:368) ~[codegen.jar:0.1]
AND MORE.. (truncated for brevity)
```
 
From what I have read on this issue (and if I understand correctly), I will not see this issue in the next release, is that correct?
**Question**
The host level connection docs http://doc.akka.io/docs/akka-http/current/scala/http/client-side/host-level.html mention that it is possible to explicitly shutdown a connection host pool, I however cannot figure out how to shutdown the specific connection pool . I my case, I would like to terminate the host connection pool when the 'responseFuture' is complete. (either successfully finished or failed after all retries). Here is my code which is very similar to the one in the docs:
```
  def requestAndParse(params: T): Future[HttpResponse] = {
  
    val urlToParse = new java.net.URL(apiUrl)
    val poolClientFlow = Http().cachedHostConnectionPool[Promise[HttpResponse]](urlToParse.getHost, urlToParse.getPort)
    val queue =
      Source.queue[(HttpRequest, Promise[HttpResponse])](QueueSize, OverflowStrategy.dropNew)
        .via(poolClientFlow)
        .toMat(Sink.foreach({
          case ((Success(resp), p)) => p.success(resp)
          case ((Failure(e), p))    => p.failure(e)
        }))(Keep.left)
        .run()

    def queueRequest(request: HttpRequest): Future[HttpResponse] = {
      val responsePromise = Promise[HttpResponse]()
      queue.offer(request -> responsePromise).flatMap {
        case QueueOfferResult.Enqueued    => responsePromise.future
        case QueueOfferResult.Dropped     => Future.failed(new RuntimeException("Queue overflowed. Try again later."))
        case QueueOfferResult.Failure(ex) =>logger.warn("Failed with " + ex)
                                            Future.failed(ex)
        case QueueOfferResult.QueueClosed => Future.failed(new RuntimeException("Queue was closed (pool shut down) while running the request. Try again later."))
      }
    }

    val request: HttpRequest = HttpRequest(
      method = HttpMethods.GET,
      uri = url(params)
    )
    // This seems relevant: https://github.com/akka/akka-http/issues/1176
    val responseFuture: Future[HttpResponse] = queueRequest(request).orTimeout(apiFutureTimeout)
    responseFuture
  }
```

**Comment by sgoyal1012** [comment]
#### Update on the question above
I managed to figure out how to obtain a reference to the host connection pool, thanks to https://stackoverflow.com/questions/34924434/akka-streams-how-to-access-the-materialized-value-of-the-stream

```
    val poolClientFlow = Http().cachedHostConnectionPool[Promise[HttpResponse]](urlToParse.getHost, urlToParse.getPort)
    val materializedObject: ((SourceQueueWithComplete[(HttpRequest, Promise[HttpResponse])], HostConnectionPool), Future[Done]) =
      Source.queue[(HttpRequest, Promise[HttpResponse])](QueueSize, OverflowStrategy.dropNew)
        .viaMat(poolClientFlow)(Keep.both)
        .toMat(Sink.foreach({
          case ((Success(resp), p)) => p.success(resp)
          case ((Failure(e), p))    => p.failure(e)
        }))(Keep.both)
        .run()


    val queue: SourceQueueWithComplete[(HttpRequest, Promise[HttpResponse])]  = materializedObject._1._1
    val pool: HostConnectionPool =  materializedObject._1._2
```

#### (Still) a Problem
This pool is used when the httpResponse future is complete, like this
```
pool.shutdown() onSuccess {
          case _ => logger.info("Shutdown the pool for My API")
        }
```
**However** I still get the same error. Maybe I need to close the connection explicitly? Because looks like closing the pool does not close the connection.

**Comment by jrudolph** [comment]
Hi @sgoyal1012, thanks for chiming in.

> From what I have read on this issue (and if I understand correctly), I will not see this stack trace in the next release, is that correct?

I think you should have never seen this exception and log unless you enabled DEBUG level logging. Have you enabled DEBUG level logging?

**Comment by jlprat** [comment]
I don't know how this happened... :thinking:

**Comment by jrudolph** [comment]
Yeah strange one.

**Comment by raboof** [comment]
The fix is in akka 2.5.3, but akka-http still depends on akka 2.4.x, so unless we backport the fix this issue will remain open until we depend on akka 2.5 (#1151)

**Comment by ktoso** [comment]
Though note that one can use Akka 2.5 already, docs how to:
https://github.com/akka/akka-http/pull/1230
But I agree about keeping ticket open until we bump the dependenc in HTTP.


On 29 June 2017 at 17:58:47, Arnout Engelen (notifications@github.com)
wrote:

The fix is in akka 2.5.3, but akka-http still depends on akka 2.4.x, so
unless we backport the fix this issue will remain open until we depend on
akka 2.5 (#1151 <https://github.com/akka/akka-http/issues/1151>)

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub
<https://github.com/akka/akka-http/issues/1176#issuecomment-311905940>, or mute
the thread

**Comment by raboof** [comment]
Jup looks like it, thanks!
